### PR TITLE
SNS-Gyroscope

### PR DESCRIPTION
**Pull request by HTTYDKing**
Separate branch for the gyroscope sensor.

### CODE REVIEW

File: lib/sensors/gyroscope.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
+#include "gyroscope.hpp"
+
+namespace hyped::sensors {
+
+Gyroscope::Gyroscope(hyped::core::ILogger &log, io::I2c &i2c, const std::uint8_t channel)
+    : log_(log),
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:6]: `logger_(logger)` for codebase consistency

File: lib/sensors/gyroscope.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+#include <cstdint>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+static constexpr std::uint8_t kGyroscope             = 0x69;
+static constexpr std::uint8_t kDataXH                = 0x29;
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.hpp:14]: `kDataXHigh`. No point shortening variable names at the expense of readability.

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.hpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <cstdint>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+static constexpr std::uint8_t kGyroscope             = 0x69;
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.hpp:13]: What is `kGyroscope`? The naming of this doesn't make the purpose of this constant clear?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.hpp (hunk 3 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.hpp:13]: `kDefaultGyroscopeAddress`

File: lib/sensors/gyroscope.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ public:
+  Gyroscope(hyped::core::ILogger &logger, io::II2c &i2c, const std::uint8_t channel);
+  ~Gyroscope();
+
+  core::Result configure();
+  std::optional<std::int16_t> read();
+  std::uint8_t getChannel();
+
+ private:
+  hyped::core::ILogger &logger_;
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.hpp:39]: redundant `hyped::`, this will cause ur checks to fail. while we are in the `hyped::xyz` namespace, there is no need to specify `hyped::` anymore, you can simply do `io::`

File: lib/sensors/gyroscope.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  }
+  const core::Result write_result5
+    = i2c_.writeByteToRegister(kGyroscope, kCtrl5, kConfigurationSetting5);
+  if (write_result5 == hyped::core::Result::kFailure) {
+    logger_.log(hyped::core::LogLevel::kFatal,
+                "Failed to, configure gyroscope control 5 at channel %d",
+                channel_);
+    return core::Result::kFailure;
+  }
+  return core::Result::kSuccess;
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:101]: dont forget success logs!

File: lib/sensors/gyroscope.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  return x_axis;
+}
+
+core::Result Gyroscope::configure()
+{
+  const core::Result write_result1
+    = i2c_.writeByteToRegister(kGyroscope, kCtrl1, kConfigurationSetting1);
+  if (write_result1 == hyped::core::Result::kFailure) {
+    logger_.log(hyped::core::LogLevel::kFatal,
+                "Failed to, configure gyroscope control 1 at channel %d",
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:73]: remove the commas after the to

File: lib/sensors/gyroscope.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+static constexpr std::uint8_t kConfigurationSetting2 = 0x20;
+static constexpr std::uint8_t kConfigurationSetting3 = 0xff;
+static constexpr std::uint8_t kConfigurationSetting5 = 0x40;
+
+class Gyroscope : public II2cMuxSensor<std::int16_t> {
+ public:
+  Gyroscope(hyped::core::ILogger &logger, io::II2c &i2c, const std::uint8_t channel);
+  ~Gyroscope();
+
+  core::Result configure();
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.hpp:34]: for this PR, I do want you to make a `static create` for gyroscope (refer to any of the I/Os for an example). in any case, I can help you with this tomorrow, but do try to do it yourself first

**Review by SnickeyX — CHANGES_REQUESTED** [review]
you're getting the hang of this

File: lib/sensors/gyroscope.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  const auto x_axis = ((gyroscope_x_high_byte.value() << 8) | gyroscope_x_low_byte.value());
+  const auto y_axis = ((gyroscope_y_high_byte.value() << 8) | gyroscope_y_low_byte.value());
+  const auto z_axis = ((gyroscope_z_high_byte.value() << 8) | gyroscope_z_low_byte.value());
+
+  logger_.log(
+    hyped::core::LogLevel::kDebug, "Successfully read from gyroscope at channel %d", channel_);
+
+  return x_axis;
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:64]: why only return `x_axis`? The function is called `read`?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/core/types.hpp [hunk 9]
```diff
@@ -9,6 +9,7 @@ static constexpr float kEpsilon = 0.0001;
 
 enum class DigitalSignal { kLow = 0, kHigh };
 enum class Result { kSuccess = 0, kFailure };
+enum class GyroscopeAxis { kX, kY, kZ };
```
Comment (SnickeyX) [inline lib/core/types.hpp:12]: make this just Axis

File: lib/sensors/gyroscope.cpp [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+{
+}
+
+std::optional<std::int16_t> Gyroscope::read(core::GyroscopeAxis axis)
+{
+  if (axis == core::GyroscopeAxis::kX) {
+    const auto gyroscope_x_high_byte = i2c_.readByte(kDefaultGyroscopeAddress, kDataXHigh);
+    if (!gyroscope_x_high_byte) {
+      logger_.log(
+        core::LogLevel::kFatal, "Failed to, read gyroscope X-axis high at channel %d", channel_);
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:22]: remove the commas!

File: lib/sensors/gyroscope.cpp [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                channel_);
+    return std::nullopt;
+  }
+}
+
+std::optional<Gyroscope> Gyroscope::create(core::ILogger &logger,
+                                           io::II2c &i2c,
+                                           const std::uint8_t channel)
+{
+  const core::Result write_result1
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:81]: be more descriptive with the name + add a comment to describe what this is even doing (i.e. what setting does it enforce?)

File: lib/sensors/gyroscope.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    = i2c.writeByteToRegister(kDefaultGyroscopeAddress, kCtrl1, kConfigurationSetting1);
+  if (write_result1 == core::Result::kFailure) {
+    logger.log(
+      core::LogLevel::kFatal, "Failed to, configure gyroscope control 1 at channel %d", channel);
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:85]: "Failed to configure the first control register for gyroscope at channel....." - same for the rest or even better would be "Failed to configure the gyroscope at channel %d with setting xyz" (xyz would be wtv the setting was)

File: lib/sensors/gyroscope.cpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    = i2c.writeByteToRegister(kDefaultGyroscopeAddress, kCtrl5, kConfigurationSetting5);
+  if (write_result5 == core::Result::kFailure) {
+    logger.log(
+      core::LogLevel::kFatal, "Failed to, configure gyroscope control 5 at channel %d", channel);
+    return std::nullopt;
+  }
+  return Gyroscope(logger, i2c, channel);
+}
+
+std::uint8_t Gyroscope::getChannel()
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.cpp:112]: `const` after getChannel

File: lib/sensors/gyroscope.hpp [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+class Gyroscope {
+ public:
+  static std::optional<Gyroscope> create(core::ILogger &logger,
+                                         io::II2c &i2c,
+                                         const std::uint8_t channel);
+  ~Gyroscope();
+
+  std::optional<std::int16_t> read(core::GyroscopeAxis axis);
+  std::uint8_t getChannel();
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.hpp:37]: `const` after getChannel

File: lib/sensors/gyroscope.hpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <cstdint>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+static constexpr std::uint8_t kDefaultGyroscopeAddress = 0x69;
```
Comment (SnickeyX) [inline lib/sensors/gyroscope.hpp:13]: mention these are from the datasheet in a comment

**Review by SnickeyX — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.cpp [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      i2c_(i2c),
+      channel_(channel)
+{
+}
+
+Gyroscope::~Gyroscope()
+{
+}
+
+std::optional<std::int16_t> Gyroscope::read(core::Axis axis)
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:16]: `const`

File: lib/sensors/gyroscope.cpp [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+{
+  if (axis == core::Axis::kX) {
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:18]: Would a switch statement not be more appropriate here?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.cpp [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+          core::LogLevel::kFatal, "Failed to read gyroscope X-axis high at channel %d", channel_);
+        return std::nullopt;
+      }
+      const auto gyroscope_x_low_byte = i2c_.readByte(kDefaultGyroscopeAddress, kDataXLow);
+      if (!gyroscope_x_low_byte) {
+        logger_.log(
+          core::LogLevel::kFatal, "Failed to read gyroscope X-axis low at channel %d", channel_);
+        return std::nullopt;
+      }
+      const auto x_axis = ((gyroscope_x_high_byte.value() << 8) | gyroscope_x_low_byte.value());
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.cpp:32]: probably just return this instead of sticking it in a variable first, same for y and z

File: lib/sensors/gyroscope.cpp [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+      return z_axis;
+    }
+  }
+}
+
+std::optional<Gyroscope> Gyroscope::create(core::ILogger &logger,
+                                           io::II2c &i2c,
+                                           const std::uint8_t channel)
+{
+  const core::Result write_result_from_Ctrl1
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.cpp:80]: `write_result_from_ctrl1`

File: lib/sensors/gyroscope.hpp [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+static constexpr std::uint8_t kDataZHigh               = 0x2D;
+static constexpr std::uint8_t kDataZLow                = 0x2C;
+static constexpr std::uint8_t kCtrl1                   = 0x20;
+static constexpr std::uint8_t kCtrl2                   = 0x21;
+static constexpr std::uint8_t kCtrl3                   = 0x22;
+static constexpr std::uint8_t kCtrl5                   = 0x24;
+static constexpr std::uint8_t kConfigurationSetting1   = 0xff;
+static constexpr std::uint8_t kConfigurationSetting2   = 0x20;
+static constexpr std::uint8_t kConfigurationSetting3   = 0xff;
+static constexpr std::uint8_t kConfigurationSetting5   = 0x40;
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.hpp:28]: comment what these configuration values are doing

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.cpp (hunk 19 above)
Comment (TomLonergan03) [reply lib/sensors/gyroscope.cpp:80]: same for the other `write_result_from_...`s

File: lib/sensors/gyroscope.hpp (hunk 20 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.hpp:28]: also remove the `static` since it isn't actually helping much with const values

File: lib/sensors/gyroscope.cpp [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+const std::optional<std::int16_t> Gyroscope::read(core::Axis axis)
+{
+  switch (axis) {
+    case core::Axis::kX: {
+      const auto gyroscope_x_high_byte = i2c_.readByte(kDefaultGyroscopeAddress, kDataXHigh);
+      if (!gyroscope_x_high_byte) {
+        logger_.log(
+          core::LogLevel::kFatal, "Failed to read gyroscope X-axis high at channel %d", channel_);
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:23]: I thought the gyroscope only mattered if the active suspension was being used? Therefore would it not make more sense to log fatal only if active suspension is being used and just logging `kWarn` if active suspension is disabled @SnickeyX @kshxtij ??

File: lib/sensors/gyroscope.hpp [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+static constexpr std::uint8_t kDataXLow                = 0x28;
+static constexpr std::uint8_t kDataYHigh               = 0x2B;
+static constexpr std::uint8_t kDataYLow                = 0x2A;
+static constexpr std::uint8_t kDataZHigh               = 0x2D;
+static constexpr std::uint8_t kDataZLow                = 0x2C;
+static constexpr std::uint8_t kCtrl1                   = 0x20;
+static constexpr std::uint8_t kCtrl2                   = 0x21;
+static constexpr std::uint8_t kCtrl3                   = 0x22;
+static constexpr std::uint8_t kCtrl5                   = 0x24;
+static constexpr std::uint8_t kConfigurationSetting1   = 0xff;
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.hpp:25]: I know we usually use full names when in doubt but `kConfigurationSetting1` seems quite verbose. Could we shorten this to `kConfigSetting1` since don't think this harms readability as it is a common abbreviation?

File: lib/sensors/gyroscope.hpp [hunk 23]
```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include "i2c_sensors.hpp"
+
+#include <cstdint>
+#include <optional>
+
+#include <core/logger.hpp>
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.hpp:8]: `#include <core/types.hpp>` This has been forward included from "i2c_sensors.hpp" but explicitly include it since you use the `Axis` enum class here and we want to keep the codebase as modular as possible.

File: lib/sensors/gyroscope.cpp (hunk 21 above)
Comment (maxguy2001) [reply lib/sensors/gyroscope.cpp:23]: This just doesn't seem like it would always be worth failing the run over...

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/debug/repl.cpp [hunk 24]
```diff
@@ -493,6 +493,60 @@ void Repl::addAccelerometerCommands(const std::uint8_t bus, const std::uint8_t d
...
+        value = gyroscope->read(core::Axis::kY);
+        break;
+      }
+      case 2: {
+        value = gyroscope->read(core::Axis::kZ);
+        break;
+      }
+      default: {
+        logger_.log(core::LogLevel::kFatal,
+                    "Failed to read recognise the gyroscope axis that was inputted from bus %d",
```
Comment (TomLonergan03) [inline lib/debug/repl.cpp:535]: "Failed to read gyroscope due to invalid axis" or similar, this is a clunky error message

File: lib/sensors/gyroscope.cpp [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+        return std::nullopt;
+      }
+      const auto gyroscope_x_low_byte = i2c_->readByte(device_address_, kDataXLow);
+      if (!gyroscope_x_low_byte) {
+        logger_.log(
+          core::LogLevel::kFatal, "Failed to read gyroscope X-axis low at channel %d", channel_);
+        return std::nullopt;
+      }
+      const auto x_axis = ((gyroscope_x_high_byte.value() << 8) | gyroscope_x_low_byte.value());
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.cpp:36]: dereference optionals using `*` not `.value()` ie `*gyroscope_x_high_byte`

File: lib/sensors/gyroscope.hpp [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+class Gyroscope {
+ public:
+  static std::optional<Gyroscope> create(core::ILogger &logger,
+                                         std::shared_ptr<io::II2c> i2c,
+                                         const std::uint8_t channel,
+                                         const std::uint8_t device_address
+                                         = kDefaultGyroscopeAddress);
+  ~Gyroscope();
+
+  const std::optional<std::int16_t> read(core::Axis axis);
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.hpp:40]: is it normal to declare return type constant? I haven't seen it elsewhere in the repo

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.cpp (hunk 18 above)
Comment (TomLonergan03) [reply lib/sensors/gyroscope.cpp:36]: nvm i get why this is like this now

File: lib/sensors/gyroscope.cpp [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+      const auto z_axis = ((*gyroscope_z_high_byte << 8) | *gyroscope_z_low_byte);
+      logger_.log(
+        core::LogLevel::kDebug, "Successfully read z-axis from gyroscope at channel %d", channel_);
+      return static_cast<std::uint16_t>(z_axis * 0.00875);
+    }
+    default: {
+      logger_.log(core::LogLevel::kFatal, "Gave an invalid axis");
+      return std::nullopt;
+    }
+  }
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.cpp:81]: i think this should probably be restructured to have a `readAxis` helper function and maybe pass in `kDataXHigh` and `kDataXLow` as parameters, right now this is a lot of almost duplicate code.

File: lib/sensors/gyroscope.cpp [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+std::optional<Gyroscope> Gyroscope::create(core::ILogger &logger,
+                                           std::shared_ptr<io::II2c> i2c,
+                                           const std::uint8_t channel,
+                                           const std::uint8_t device_address)
+{
+  const core::Result write_result_from_ctrl1
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.cpp:89]: `auto` for all the `writeByte` calls

File: lib/sensors/gyroscope.hpp [hunk 29]
```diff
@@ -0,0 +1,56 @@
...
+constexpr std::uint8_t kConfigurationSetting3   = 0xff;
+constexpr std::uint8_t kConfigurationSetting5   = 0x40;
+
+class Gyroscope {
+ public:
+  static std::optional<Gyroscope> create(core::ILogger &logger,
+                                         std::shared_ptr<io::II2c> i2c,
+                                         const std::uint8_t channel,
+                                         const std::uint8_t device_address
+                                         = kDefaultGyroscopeAddress);
```
Comment (TomLonergan03) [inline lib/sensors/gyroscope.hpp:38]: i am not keen on having default parameters, if a gyro will always be at `kDefaultGyroscopeAddress` then rename to `kGyroscopeAddress` and drop the `device_address` parameter and instead use the constant, or if a gyro might have other addresses then the address should always be passed in when constructing.

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.cpp (hunk 21 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.cpp:27]: thats an amazing catch, going to push for my overhaul to get merged to we can merge master and get `kWarn` here

File: lib/sensors/gyroscope.hpp (hunk 29 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.hpp:38]: it can have two different addresses, @HTTYDKing make sure you add the "alternative address" as well, think its 0x68

File: lib/sensors/gyroscope.hpp (hunk 26 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.hpp:41]: this actually makes sense since this just enforeces that member variables arent modifed; although the `const` should be suffixed

File: lib/debug/repl.cpp [hunk 30]
```diff
@@ -493,6 +493,59 @@ void Repl::addAccelerometerCommands(const std::uint8_t bus, const std::uint8_t d
...
+  std::stringstream description;
+  description << "Read gyroscope sensor 0x" << std::hex << static_cast<int>(device_address)
+              << " on "
+              << "I2C bus " << static_cast<int>(bus);
+  gyroscope_read_command.description = description.str();
+  gyroscope_read_command.handler     = [this, gyroscope, bus]() {
+    std::uint16_t axis;
+    std::cout << "axis (0 is x, 1 is y, 2 is z): ";
+    std::cin >> axis;
+    std::optional<std::int16_t> value;
```
Comment (maxguy2001) [inline lib/debug/repl.cpp:519]: `auto`?

File: lib/sensors/gyroscope.hpp [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+constexpr std::uint8_t kCtrl3                   = 0x22;
+constexpr std::uint8_t kCtrl5                   = 0x24;
+constexpr std::uint8_t kConfigurationSetting1   = 0xff;
+constexpr std::uint8_t kConfigurationSetting2   = 0x20;
+constexpr std::uint8_t kConfigurationSetting3   = 0xff;
+constexpr std::uint8_t kConfigurationSetting5   = 0x40;
+
+class Gyroscope {
+ public:
+  static std::optional<Gyroscope> create(core::ILogger &logger,
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.hpp:34]: Do we still explicitly call `static`? @SnickeyX

File: lib/sensors/gyroscope.cpp [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+      channel_(channel),
+      device_address_(device_address)
+{
+}
+
+Gyroscope::~Gyroscope()
+{
+}
+
+const std::optional<std::int16_t> Gyroscope::read(core::Axis axis)
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:20]: Does `const` not go after instead of before?

File: lib/sensors/gyroscope.cpp [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+      const auto gyroscope_x_low_byte = i2c_->readByte(device_address_, kDataXLow);
+      if (!gyroscope_x_low_byte) {
+        logger_.log(
+          core::LogLevel::kFatal, "Failed to read gyroscope X-axis low at channel %d", channel_);
+        return std::nullopt;
+      }
+      const auto x_axis = ((*gyroscope_x_high_byte << 8) | *gyroscope_x_low_byte);
+      logger_.log(
+        core::LogLevel::kDebug, "Successfully read x-axis from gyroscope at channel %d", channel_);
+      return static_cast<std::uint16_t>(x_axis * 0.00875);
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:39]: What is 0.00875 and where did it come from?

Comment (maxguy2001) [reply lib/sensors/gyroscope.cpp:39]: Also elsewhere in the file. Maybe give it a name and store it somewhere?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/sensors/gyroscope.hpp (hunk 31 above)
Comment (SnickeyX) [reply lib/sensors/gyroscope.hpp:34]: yes here its necessary, else `create` will not be able to access gyroscope specifc functions

**Comment by SnickeyX** [comment]
@HTTYDKing remember to merge master for merge conflicts

File: lib/sensors/gyroscope.cpp (hunk 28 above)
Comment (TomLonergan03) [reply lib/sensors/gyroscope.cpp:89]: outdated, leave as is

File: lib/core/types.hpp [hunk 34]
```diff
@@ -62,6 +62,18 @@ struct RawAccelerationData {
   const TimePoint measured_at;
 };
 
+struct GyroscopeData {
```
Comment (maxguy2001) [inline lib/core/types.hpp:65]: What's the use for this? Is there no more suitable alternatives?

File: lib/debug/repl.cpp [hunk 35]
```diff
@@ -196,6 +196,28 @@ std::optional<std::unique_ptr<Repl>> Repl::fromFile(const std::string &path)
...
+                "Missing required field 'sensors.gyroscope.enabled' in configuration file");
+    return std::nullopt;
+  }
+  if (gyroscope["enabled"].GetBool()) {
+    if (!gyroscope.HasMember("bus")) {
+      logger_.log(core::LogLevel::kFatal,
+                  "Missing required field 'sensors.gyroscope.bus' in configuration file");
+      return std::nullopt;
+    }
+    const auto device_address = gyroscope["device_address"].GetUint();
```
Comment (maxguy2001) [inline lib/debug/repl.cpp:217]: Did you check this exists first?

File: lib/debug/repl.cpp [hunk 36]
```diff
@@ -493,6 +515,53 @@ void Repl::addAccelerometerCommands(const std::uint8_t bus, const std::uint8_t d
...
+    auto Old_time = std::chrono::system_clock::now();
+    std::optional<core::GyroscopeData> value;
+    
+    while (true) {
+      std::optional<core::GyroscopeData> value = gyroscope->read(core::Axis::kX);
+      const auto result = value.value();
+
+      std::chrono::duration<double> elapsed_seconds = result.measured_at - Old_time;
+      
+      auto time = (core::Float) elapsed_seconds.count();
```
[the remark this answers is not in the document]
Comment (maxguy2001) [reply lib/debug/repl.cpp:549]: also, more informative naming?

File: lib/debug/repl.cpp [hunk 37]
```diff
@@ -493,6 +515,53 @@ void Repl::addAccelerometerCommands(const std::uint8_t bus, const std::uint8_t d
...
+
+      Angle = 0.5*(Old_data+result.x)*time + Angle;
```
Comment (maxguy2001) [inline lib/debug/repl.cpp:551]: `angle`

Comment (maxguy2001) [reply lib/debug/repl.cpp:551]: Also where does this function come from?

Comment (maxguy2001) [reply lib/debug/repl.cpp:551]: and you can just use `angle += ...`

File: lib/debug/repl.hpp [hunk 38]
```diff
@@ -14,6 +14,9 @@
 #include <io/hardware_uart.hpp>
 #include <io/pwm.hpp>
 #include <sensors/accelerometer.hpp>
+#include <sensors/gyroscope.hpp>
+#include <unistd.h>
+#include <chrono>
```
Comment (maxguy2001) [inline lib/debug/repl.hpp:19]: Is this used?

File: lib/sensors/gyroscope.cpp [hunk 39]
```diff
@@ -0,0 +1,146 @@
...
+std::optional<Gyroscope> Gyroscope::create(core::ILogger &logger,
+                                           std::shared_ptr<io::II2c> i2c,
+                                           const std::uint8_t channel,
+                                           const std::uint8_t device_address)
+{
+  const core::Result write_result_from_ctrl1
+    = i2c->writeByteToRegister(device_address, kCtrl1, kConfigurationSetting1);
+  if (write_result_from_ctrl1 == core::Result::kFailure) {
+    logger.log(
+      core::LogLevel::kFatal,
```
Comment (maxguy2001) [inline lib/sensors/gyroscope.cpp:108]: Again, isn't this only critical failure if we're actually using active suspension?

File: lib/core/types.hpp (hunk 34 above)
Comment (maxguy2001) [reply lib/core/types.hpp:65]: Shouldn't this be a static thing that can get passed somewhere for full processing?

File: lib/debug/repl.cpp [hunk 40]
```diff
@@ -493,6 +515,53 @@ void Repl::addAccelerometerCommands(const std::uint8_t bus, const std::uint8_t d
...
+    core::Float Angle = 0;
+    core::Float Old_data = 0;
+    auto Old_time = std::chrono::system_clock::now();
+    std::optional<core::GyroscopeData> value;
+    
+    while (true) {
+      std::optional<core::GyroscopeData> value = gyroscope->read(core::Axis::kX);
+      const auto result = value.value();
+
+      std::chrono::duration<double> elapsed_seconds = result.measured_at - Old_time;
```
Comment (maxguy2001) [inline lib/debug/repl.cpp:547]: Have a look in core. Most of that functionality should already be implemented and if possible, we want to keep the ugly chrono notation contained in there.

**Review by maxguy2001 — CHANGES_REQUESTED** [review]
